python/nvtx/nvtx.py

```diff
@@ -0,0 +1,146 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.
+
+from contextlib import ContextDecorator
+
+from nvtx._lib import (
+    Domain,
+    EventAttributes,
+    mark as libnvtx_mark,
+    pop_range as libnvtx_pop_range,
+    push_range as libnvtx_push_range,
+)
+
+
+class annotate(ContextDecorator):
+    """
+    Annotate code ranges using a context manager or a decorator.
+    """
+
+    def __init__(self, message=None, color=None, domain=None):
+        """
+        Annotate a function or a code range.
+
+        Parameters
+        ----------
+        message : str, optional
+            A message associated with the annotated code range.
+            When used as a decorator, the default value of message
+            is the name of the function being decorated.
+            When used as a context manager, the default value
+            is the empty string.
+        color : str or color, optional
+            A color associated with the annotated code range.
+            Supports `matplotlib` colors if it is available.
+        domain : str, optional
+            Name of a domain under which the code range is scoped.
+            The default domain name is "NVTX".
+
+        Examples
+        --------
+        >>> import nvtx
+        >>> import time
+
+        Using a decorator:
+
+        >>> @nvtx.annotate("my_func", color="red", domain="my_domain")
+        ... def func():
+        ...     time.sleep(0.1)
+
+        Using a context manager:
+
+        >>> with nvtx.annotate("my_code_range", color="blue"):
+        ...    time.sleep(10)
+        ...
+        """
+        self.attributes = EventAttributes(message, color)
+        self.domain = Domain(domain)
+
+    def __reduce__(self):
+        return (
+            self.__class__,
+            (self.attributes.message, self.attributes.color, self.domain.name),
+        )
+
+    def __enter__(self):
+        libnvtx_push_range(self.attributes, self.domain.handle)
+        return self
+
+    def __exit__(self, *exc):
+        libnvtx_pop_range(self.domain.handle)
+        return False
+
+    def __call__(self, func):
+        if not self.attributes.message:
+            self.attributes.message = func.__name__
+        return super().__call__(func)
+
+
+class _annotate_nop:
+    def __init__(self, *args, **kwargs):
+        pass
+
+    def __enter__(self):
+        pass
+
+    def __exit__(self, *args, **kwargs):
+        return False
+
+    def __call__(self, func):
+        return func
```

Comment:
If we're keeping the no-op env variable, may want to copy the doc from `annotate` on this so the user can still get a nice docstring even if the env variable turns it to a no-op.

```suggestion
class _annotate_nop:
    def __init__(self, *args, **kwargs):
        pass

    def __enter__(self):
        pass

    def __exit__(self, *args, **kwargs):
        return False

    def __call__(self, func):
        return func


_annotate_nop.__doc__ = annotate.__doc__
_annotate_nop.__init__.__doc__ = annotate.__init__.__doc__
```

Reply:
Just got rid of the no-op stuff for now